active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,959 @@
+      "fileOverrides": {
+        "*.stack.bicep": "error",
+        "<glob>": "error"
+      }
+    }
+  }
+}
+```
+
+There will be no ARM template changes to drive this. The extra validation on the backend will take place if the
+deployment is sent from the Stacks service. The Bicep compilation step for Stacks CLI commands should run with this mode
+set to "error" regardless of user setting to prevent templates that will certainly fail from being sent to the API.
+
+#### Passing extension configurations to nested deployments
+
+Users will need to reuse extension configurations across multiple nested deployments. Because these are passed as
+deployment properties rather than as parameters, there needs to be a way to reference this configuration in the
+consuming deployment for the nested deployment.
+
+Here are some examples of passing a configuration down another level:
+
+```bicep
+extension kubernetes // this was provided a configuration through the parent deployment/parameters file
+
+// ✅ whole inheritance
+module foo 'foo.bicep' = {
+  extensionConfigs: {
+    kubernetes: kubernetes // resuse its config, must be from an extension with the same name and version
+  }
+}
+
+// ✅ piecemeal inheritance
+module baz 'foo.bicep' = {
+  extensionConfigs: {
+    kubernetes: { 
+      kubeConfig: kubernetes.kubeConfig // types must match
+      namespace: kubernetes.namespace // types must match
+    }
+  }
+}
+
+// ❌ spread inheritance CANNOT be used
+module bar 'foo.bicep' = {
+  extensionConfigs: {
+    kubernetes: {
+       ...kubernetes
+       namespace: 'myOtherNamespace'
+    }
+  }
+}
+
+// ❌ secure properties cannot be used as values to non-secure properties
+module fooy 'foo.bicep' = {
+  extensionConfigs: {
+    kubernetes: { 
+      kubeConfig: kubernetes.kubeConfig // ok
+      namespace: kubernetes.kubeConfig // error
+    }
+  }
+}
+```
+
+##### Limitations
+
+Because secure property values must be provided in reference format, there are several limitations on allowable Bicep
+expressions inside of extension configurations for any deployment:
+
+1. No object spreads within an expression scope that is not represented as a language expression in the ARM template.
+  This includes the top-level extension configuration object and secure property reference objects.
+1. Extension configuration inheritance is limited to whole inheritance or explicit piecemeal inheritance (each property
+  must be explicitly written out to inherit from some value).
+
+#### Providing extension configurations to root deployments
+
+Because extension configurations will be an input property to the deployment, there needs to be a way for frontends
+to supply this data to a root deployment. The `bicepparam` file type can be expanded to support defining extension
+configurations.
+
+Here is an example:
+
+main.bicepparam
+
+```bicep
+using 'main.bicep'
+
+param boolParam = true
+// ...
+
+extensionConfigs {
```

Comment:
An alternative approach would be to reuse the `extension` and `with` keywords for supplying extension configurations, using a separate statement for each extension:

```
extension k8s with {
  namespace: 'default'
  ...
}
```

Reply:
To clarify, this isn't a suggestion. I'm simply thinking out loud and offering an alternative.

Reply:
I agree this is better for the params file. Will update.